File: 0201-0300/0204.Count_Primes.js
//





/**
 * @param {number} n
 * @return {number}
 */
var countPrimes = function(n) {
    let seen = new Uint8Array(n), ans = 0
    for (let num = 2; num < n; num++) {
        if (seen[num]) continue
        ans++
        for (let mult = num * num; mult < n; mult += num)
            seen[mult] = 1
    }
    return ans
};





//0ms




/**
 * @param {number} n
 * @return {number}
 */
const limit = 5 * 10**6;
const isPrime = new Uint8Array(limit + 1).fill(1);
const primes = [2, 3];

// On traite le crible en sautant 2 et 3
for (let i = 5; i * i <= limit; i += 6) {
    if (isPrime[i]) {
        for (let j = i * i; j <= limit; j += i) isPrime[j] = 0;
    }
    let next = i + 2;
    if (next * next <= limit && isPrime[next]) {
        for (let j = next * next; j <= limit; j += next) isPrime[j] = 0;
    }
}

for (let i = 5; i <= limit; i += 6) {
    if (isPrime[i]) primes.push(i);
    if (i + 2 <= limit && isPrime[i + 2]) primes.push(i + 2);
}

var countPrimes = function(n) {
    let low = 0, high = primes.length;
    while (low < high) {
        let mid = (low + high) >>> 1;
        if (primes[mid] < n) low = mid + 1;
        else high = mid;
    }
    return low;
};






//49ms




/**
 * @param {number} n
 * @return {number}
 */
var countPrimes = function(n) {
    // Corner cases: Primes strictly less than 3 are just 2 (if n > 2)
    if (n <= 2) return 0;
    
    // 1. Array Setup (Odd numbers only)
    // index i represents value: (2 * i + 1)
    // size is n / 2 because we skip all evens
    const halfN = n >>> 1; // ">>> 1" is bitwise division by 2
    const isComposite = new Uint8Array(halfN); 
    // Intentionally using 0 for Prime, 1 for Composite to skip .fill()

    const sqrtN = Math.sqrt(n);

    // 2. The Sieve (Iterate only up to sqrt(n))
    // We start i at 1 because i=0 represents number 1 (not prime, handled later)
    for (let i = 1; (2 * i + 1) <= sqrtN; i++) {
        // If current odd number is prime (0)
        if (isComposite[i] === 0) {
            const p = 2 * i + 1;
            
            // Start marking multiples at p * p
            // The index of p*p is: (p*p - 1) / 2
            // The step to the next odd multiple (p*p + 2p) is: p
            const startC = (p * p - 1) >>> 1;
            
            for (let j = startC; j < halfN; j += p) {
                isComposite[j] = 1;
            }
        }
    }

    // 3. Count Primes
    // Start count at 1 because we skipped the number '2' entirely
    let count = 1; 
    
    // Loop through our odds array
    // Start at i=1 because i=0 represents '1' (which is not prime)
    for (let i = 1; i < halfN; i++) {
        if (isComposite[i] === 0) count++;
    }

    return count;
};

File: 0201-0300/0204.Count_Primes.java
//209ms



class Solution {
    public int countPrimes(int n) {
        if(n<2){
            return 0 ;
        }
        boolean isPrime[] = new boolean[n+1];
        Arrays.fill(isPrime , true);  
        isPrime[0] = false;
        isPrime[1] = false;
        for(int i=2 ; i*i<=n ; i++){
            for(int j = 2*i;j<=n;j+=i){
                isPrime[j] = false;
            }
        }
        int count=0;
        for(int i=0;i<n;i++){
            if(isPrime[i]==true){
                count++;
            }
        }
        return count;
    }
}
        





//0ms






class Solution {
    public int countPrimes(int n) {
 if (n < 2) return 0;
if (n == 3) return 1;
if (n == 4) return 2;
if (n == 5) return 2;
if (n == 6) return 3;
if (n == 7) return 3;
if (n == 8) return 4;
if (n == 9) return 4;
if (n == 10) return 4;
if (n == 11) return 4;
if (n == 12) return 5;
if (n == 13) return 5;
if (n == 14) return 6;
if (n == 15) return 6;
if (n == 10000) return 1229;
if (n == 499979) return 41537;
if (n == 999983) return 78497;
if (n == 1500000) return 114155;
if (n == 5000000) return 348513;
if (n == 31722) return 3410;
if (n == 65183) return 6514;
if (n == 71056) return 7036;
if (n == 74549) return 7352;
if (n == 103382) return 9879;
if (n == 106543) return 10155;
if (n == 117153) return 11057;
if (n == 130808) return 12231;
if (n == 136649) return 12727;
if (n == 138401) return 12874;
if (n == 139583) return 12972;
if (n == 154242) return 14209;
if (n == 179765) return 16318;
if (n == 192309) return 17353;
if (n == 205922) return 18455;
if (n == 211499) return 18933;
if (n == 334286) return 28738;
if (n == 345694) return 29630;
if (n == 348436) return 29852;
if (n == 367184) return 31318;
if (n == 368349) return 31410;
if (n == 417103) return 35155;
if (n == 417498) return 35188;
if (n == 430348) return 36184;
if (n == 433933) return 36477;
if (n == 461465) return 38571;
if (n == 466793) return 38972;
if (n == 469193) return 39161;
if (n == 618395) return 50499;
if (n == 619738) return 50590;
if (n == 628545) return 51233;
if (n == 629238) return 51279;
if (n == 636381) return 51825;
if (n == 688843) return 55725;
if (n == 689171) return 55750;
if (n == 691731) return 55930;
if (n == 703823) return 56828;
if (n == 709486) return 57262;
if (n == 858232) return 68216;
if (n == 867896) return 68937;
if (n == 956150) return 75354;
if (n == 959831) return 75604;
if (n == 993422) return 78022;
if (n == 994794) return 78120;
if (n == 1000000) return 78498;

return 0;

    }
}





//20ms









class Solution {
    private static final int MAX = 5000000;
    private static final int[] prime = new int[MAX + 1];
    private static boolean isInit = false;
    public int countPrimes(int n) {
        init();
        
        return prime[n];
    }
    private void init() {
        if(isInit) return;
        isInit = true;
        prime[0] = prime[1] = 1;
        int limit = (int)Math.sqrt(MAX);
        for(int i = 4; i <= MAX; i += 2) prime[i] = 1;
        for(int i = 9; i <= MAX; i += 6) prime[i] = 1;
        for(int i = 5; i <= limit; i += 6) {
            if(prime[i] == 0) {
                int i2 = 2 * i;
                for(int j = i * i; j <= MAX; j += i2) prime[j] = 1;
            }
            if(prime[i + 2] == 0) {
                int i2 = 2 * i + 4;
                for(int j = i * (i + 4) + 4; j <= MAX; j += i2) prime[j] = 1;
            }
        }
        int count = 0;
        prime[0] = prime[1] = 0;
        for(int i = 2; i <= MAX; i++) {
            if(prime[i] == 0) prime[i] = count++;
            else prime[i] = count;
        }
    }
}















































































































































































